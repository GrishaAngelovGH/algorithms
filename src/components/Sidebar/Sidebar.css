.pro-sidebar {
    height: auto;
    width: 220px;
    min-width: 225px;
}

.pro-sidebar > .pro-sidebar-inner {
    background: #ffffff;
    color: #000000;
}

.pro-sidebar .pro-menu > ul > .pro-sub-menu > .pro-inner-list-item {    
    background-color: #ffffff;
}

.pro-sidebar .pro-menu a {
    color:#000000;
}