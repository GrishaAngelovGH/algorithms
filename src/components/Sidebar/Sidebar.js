import { useState } from 'react'
import { Link } from 'react-router-dom'

import {
    ProSidebar, SidebarHeader,
    Menu, MenuItem, SubMenu
} from 'react-pro-sidebar'

import 'react-pro-sidebar/dist/css/styles.css'
import './Sidebar.css'

const SubMenuItem = ({ id, open, subMenuTitle, menuItems, onOpenChange }) => {

    const handleOpenChange = () => {
        onOpenChange(id)
    }

    return (
        <SubMenu
            title={subMenuTitle}
            open={open}
            className='bg-primary text-white'
            onOpenChange={handleOpenChange}
        >
            {
                menuItems.map(v => (
                    <MenuItem key={v.title}>
                        <Link to={v.link}>{v.title}</Link>
                    </MenuItem>
                ))
            }
        </SubMenu>
    )
}

const Sidebar = () => {
    const [menu, setMenu] = useState([
        {
            id: 1,
            subMenuTitle: 'Searching',
            open: true,
            menuItems: [
                { link: '/linear-search', title: 'Linear Search' },
                { link: '/binary-search', title: 'Binary Search' }
            ]
        },
        {
            id: 2,
            subMenuTitle: 'Sorting',
            open: false,
            menuItems: [
                { link: '/bubble-sort', title: 'Bubble Sort' },
                { link: '/quick-sort', title: 'Quick Sort' }
            ]
        },
        {
            id: 3,
            subMenuTitle: 'Binary Tree',
            open: false,
            menuItems: [
                { link: '/binary-tree', title: 'BFS and DFS Traversal' }
            ]
        },
        {
            id: 4,
            subMenuTitle: 'Cryptography',
            open: false,
            menuItems: [
                { link: '/caesar-cipher', title: 'Caesar Cipher' }
            ]
        },
        {
            id: 5,
            subMenuTitle: 'Computer Graphics',
            open: false,
            menuItems: [
                { link: '/flood-fill', title: 'Flood Fill' }
            ]
        }
    ])

    const handleOpenChange = id => {
        const newMenu = menu.map(v => {
            if (v.id === id) {
                return {
                    ...v,
                    open: true
                }
            }

            return {
                ...v,
                open: false
            }
        })

        setMenu(newMenu)
    }

    return (
        <ProSidebar>
            <SidebarHeader>
                <h3 className='text-white bg-primary text-center p-2'>Algorithms</h3>
            </SidebarHeader>

            <Menu iconShape='square'>
                {
                    menu.map(subMenu => (
                        <SubMenuItem key={subMenu.subMenuTitle} {...subMenu} onOpenChange={handleOpenChange} />
                    ))
                }
            </Menu>
        </ProSidebar>
    )
}

export default Sidebar