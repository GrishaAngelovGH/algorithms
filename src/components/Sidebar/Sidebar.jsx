import { useState } from 'react'
import { Link } from 'react-router-dom'
import { Sidebar, Menu, SubMenu } from 'react-pro-sidebar'

import './Sidebar.css'

export const SubMenuWrapper = ({ subMenuTitle, open, menuItems }) => {
	const [isOpen, setIsOpen] = useState(open)

	const handleOpenChange = openValue => {
		setIsOpen(openValue)
	}

	return (
		<SubMenu label={subMenuTitle} open={isOpen} onOpenChange={handleOpenChange}>
			{
				menuItems.map((v, i) => (
					<Link key={i} className='d-block text-center p-1 text-decoration-none' to={v.link}>
						{v.title}
					</Link>
				))
			}
		</SubMenu>
	)
}

const SidebarWrapper = () => {
	const menu = [
		{
			id: 1,
			subMenuTitle: 'Searching',
			open: true,
			menuItems: [
				{ link: '/linear-search', title: 'Linear Search' },
				{ link: '/binary-search', title: 'Binary Search' }
			]
		},
		{
			id: 2,
			subMenuTitle: 'Sorting',
			open: false,
			menuItems: [
				{ link: '/bubble-sort', title: 'Bubble Sort' },
				{ link: '/quick-sort', title: 'Quick Sort' }
			]
		},
		{
			id: 3,
			subMenuTitle: 'Binary Tree',
			open: false,
			menuItems: [
				{ link: '/binary-tree', title: 'BFS and DFS Traversal' }
			]
		},
		{
			id: 4,
			subMenuTitle: 'Cryptography',
			open: false,
			menuItems: [
				{ link: '/caesar-cipher', title: 'Caesar Cipher' }
			]
		},
		{
			id: 5,
			subMenuTitle: 'Computer Graphics',
			open: false,
			menuItems: [
				{ link: '/flood-fill', title: 'Flood Fill' }
			]
		}
	]

	return (
		<Sidebar width='230px' style={{ zIndex: 100 }} backgroundColor='dodgerblue'>
			<h3 className='text-white bg-primary text-center p-2'>Algorithms</h3>
			<Menu>
				{
					menu.map(v => (
						<SubMenuWrapper
							key={v.id}
							open={v.open}
							menuItems={v.menuItems}
							subMenuTitle={v.subMenuTitle}
						/>
					))
				}
			</Menu>
		</Sidebar>
	)
}

export default SidebarWrapper