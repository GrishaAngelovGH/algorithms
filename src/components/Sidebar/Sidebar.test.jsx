import {
	BrowserRouter as Router,
	Routes,
	Route
} from 'react-router-dom'

import { render } from '@testing-library/react'

import { Menu } from 'react-pro-sidebar'
import Sidebar, { SubMenuWrapper } from './Sidebar'

test('should render Sidebar component', () => {
	const view = render(
		<Router>
			<Routes>
				<Route
					path='/'
					element={(
						<Sidebar />
					)}
				/>
			</Routes>
		</Router>
	)

	expect(view).toMatchSnapshot()
})

test('should render SubMenu component', () => {
	const subMenu = {
		subMenuTitle: 'SubMenuTitle',
		open: false,
		menuItems: [
			{ link: '/', title: 'Title' }
		]
	}

	const view = render(
		<Router>
			<Routes>
				<Route
					path='/'
					element={(
						<Menu>
							<SubMenuWrapper
								open={subMenu.open}
								menuItems={subMenu.menuItems}
								subMenuTitle={subMenu.subMenuTitle}
							/>
						</Menu>
					)}
				/>
			</Routes>
		</Router>
	)

	expect(view).toMatchSnapshot()
})