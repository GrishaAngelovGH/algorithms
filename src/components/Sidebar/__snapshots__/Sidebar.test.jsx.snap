// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render Sidebar component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <aside
        class="ps-sidebar-root css-1gjb6qj"
        data-testid="ps-sidebar-root-test-id"
        style="z-index: 100;"
        width="230px"
      >
        <div
          class="ps-sidebar-container css-1o85esu"
          data-testid="ps-sidebar-container-test-id"
        >
          <h3
            class="text-white bg-primary text-center p-2"
          >
            Algorithms
          </h3>
          <nav
            class="ps-menu-root css-vj11vy"
          >
            <ul
              class="css-ewdv3l"
            >
              <li
                class="ps-menuitem-root ps-submenu-root ps-open css-16jesut"
              >
                <a
                  class="ps-menu-button ps-open"
                  data-testid="ps-menu-button-test-id"
                  tabindex="0"
                >
                  <span
                    class="ps-menu-label ps-open css-12w9als"
                  >
                    Searching
                  </span>
                  <span
                    class="ps-submenu-expand-icon ps-open css-1cuxlhl"
                  >
                    <span
                      class="css-jn69v9"
                      open=""
                    />
                  </span>
                </a>
                <div
                  class="ps-submenu-content ps-open css-nx2aea"
                  data-testid="ps-submenu-content-test-id"
                  open=""
                >
                  <ul
                    class="css-ewdv3l"
                  >
                    <a
                      class="d-block text-center p-1 text-decoration-none"
                      data-discover="true"
                      href="/linear-search"
                    >
                      Linear Search
                    </a>
                    <a
                      class="d-block text-center p-1 text-decoration-none"
                      data-discover="true"
                      href="/binary-search"
                    >
                      Binary Search
                    </a>
                  </ul>
                </div>
              </li>
              <li
                class="ps-menuitem-root ps-submenu-root css-16jesut"
              >
                <a
                  class="ps-menu-button"
                  data-testid="ps-menu-button-test-id"
                  tabindex="0"
                >
                  <span
                    class="ps-menu-label css-12w9als"
                  >
                    Sorting
                  </span>
                  <span
                    class="ps-submenu-expand-icon css-1cuxlhl"
                  >
                    <span
                      class="css-honxw6"
                    />
                  </span>
                </a>
                <div
                  class="ps-submenu-content css-18unl23"
                  data-testid="ps-submenu-content-test-id"
                >
                  <ul
                    class="css-ewdv3l"
                  >
                    <a
                      class="d-block text-center p-1 text-decoration-none"
                      data-discover="true"
                      href="/bubble-sort"
                    >
                      Bubble Sort
                    </a>
                    <a
                      class="d-block text-center p-1 text-decoration-none"
                      data-discover="true"
                      href="/quick-sort"
                    >
                      Quick Sort
                    </a>
                  </ul>
                </div>
              </li>
              <li
                class="ps-menuitem-root ps-submenu-root css-16jesut"
              >
                <a
                  class="ps-menu-button"
                  data-testid="ps-menu-button-test-id"
                  tabindex="0"
                >
                  <span
                    class="ps-menu-label css-12w9als"
                  >
                    Binary Tree
                  </span>
                  <span
                    class="ps-submenu-expand-icon css-1cuxlhl"
                  >
                    <span
                      class="css-honxw6"
                    />
                  </span>
                </a>
                <div
                  class="ps-submenu-content css-18unl23"
                  data-testid="ps-submenu-content-test-id"
                >
                  <ul
                    class="css-ewdv3l"
                  >
                    <a
                      class="d-block text-center p-1 text-decoration-none"
                      data-discover="true"
                      href="/binary-tree"
                    >
                      BFS and DFS Traversal
                    </a>
                  </ul>
                </div>
              </li>
              <li
                class="ps-menuitem-root ps-submenu-root css-16jesut"
              >
                <a
                  class="ps-menu-button"
                  data-testid="ps-menu-button-test-id"
                  tabindex="0"
                >
                  <span
                    class="ps-menu-label css-12w9als"
                  >
                    Cryptography
                  </span>
                  <span
                    class="ps-submenu-expand-icon css-1cuxlhl"
                  >
                    <span
                      class="css-honxw6"
                    />
                  </span>
                </a>
                <div
                  class="ps-submenu-content css-18unl23"
                  data-testid="ps-submenu-content-test-id"
                >
                  <ul
                    class="css-ewdv3l"
                  >
                    <a
                      class="d-block text-center p-1 text-decoration-none"
                      data-discover="true"
                      href="/caesar-cipher"
                    >
                      Caesar Cipher
                    </a>
                  </ul>
                </div>
              </li>
              <li
                class="ps-menuitem-root ps-submenu-root css-16jesut"
              >
                <a
                  class="ps-menu-button"
                  data-testid="ps-menu-button-test-id"
                  tabindex="0"
                >
                  <span
                    class="ps-menu-label css-12w9als"
                  >
                    Computer Graphics
                  </span>
                  <span
                    class="ps-submenu-expand-icon css-1cuxlhl"
                  >
                    <span
                      class="css-honxw6"
                    />
                  </span>
                </a>
                <div
                  class="ps-submenu-content css-18unl23"
                  data-testid="ps-submenu-content-test-id"
                >
                  <ul
                    class="css-ewdv3l"
                  >
                    <a
                      class="d-block text-center p-1 text-decoration-none"
                      data-discover="true"
                      href="/flood-fill"
                    >
                      Flood Fill
                    </a>
                  </ul>
                </div>
              </li>
            </ul>
          </nav>
        </div>
      </aside>
    </div>
  </body>,
  "container": <div>
    <aside
      class="ps-sidebar-root css-1gjb6qj"
      data-testid="ps-sidebar-root-test-id"
      style="z-index: 100;"
      width="230px"
    >
      <div
        class="ps-sidebar-container css-1o85esu"
        data-testid="ps-sidebar-container-test-id"
      >
        <h3
          class="text-white bg-primary text-center p-2"
        >
          Algorithms
        </h3>
        <nav
          class="ps-menu-root css-vj11vy"
        >
          <ul
            class="css-ewdv3l"
          >
            <li
              class="ps-menuitem-root ps-submenu-root ps-open css-16jesut"
            >
              <a
                class="ps-menu-button ps-open"
                data-testid="ps-menu-button-test-id"
                tabindex="0"
              >
                <span
                  class="ps-menu-label ps-open css-12w9als"
                >
                  Searching
                </span>
                <span
                  class="ps-submenu-expand-icon ps-open css-1cuxlhl"
                >
                  <span
                    class="css-jn69v9"
                    open=""
                  />
                </span>
              </a>
              <div
                class="ps-submenu-content ps-open css-nx2aea"
                data-testid="ps-submenu-content-test-id"
                open=""
              >
                <ul
                  class="css-ewdv3l"
                >
                  <a
                    class="d-block text-center p-1 text-decoration-none"
                    data-discover="true"
                    href="/linear-search"
                  >
                    Linear Search
                  </a>
                  <a
                    class="d-block text-center p-1 text-decoration-none"
                    data-discover="true"
                    href="/binary-search"
                  >
                    Binary Search
                  </a>
                </ul>
              </div>
            </li>
            <li
              class="ps-menuitem-root ps-submenu-root css-16jesut"
            >
              <a
                class="ps-menu-button"
                data-testid="ps-menu-button-test-id"
                tabindex="0"
              >
                <span
                  class="ps-menu-label css-12w9als"
                >
                  Sorting
                </span>
                <span
                  class="ps-submenu-expand-icon css-1cuxlhl"
                >
                  <span
                    class="css-honxw6"
                  />
                </span>
              </a>
              <div
                class="ps-submenu-content css-18unl23"
                data-testid="ps-submenu-content-test-id"
              >
                <ul
                  class="css-ewdv3l"
                >
                  <a
                    class="d-block text-center p-1 text-decoration-none"
                    data-discover="true"
                    href="/bubble-sort"
                  >
                    Bubble Sort
                  </a>
                  <a
                    class="d-block text-center p-1 text-decoration-none"
                    data-discover="true"
                    href="/quick-sort"
                  >
                    Quick Sort
                  </a>
                </ul>
              </div>
            </li>
            <li
              class="ps-menuitem-root ps-submenu-root css-16jesut"
            >
              <a
                class="ps-menu-button"
                data-testid="ps-menu-button-test-id"
                tabindex="0"
              >
                <span
                  class="ps-menu-label css-12w9als"
                >
                  Binary Tree
                </span>
                <span
                  class="ps-submenu-expand-icon css-1cuxlhl"
                >
                  <span
                    class="css-honxw6"
                  />
                </span>
              </a>
              <div
                class="ps-submenu-content css-18unl23"
                data-testid="ps-submenu-content-test-id"
              >
                <ul
                  class="css-ewdv3l"
                >
                  <a
                    class="d-block text-center p-1 text-decoration-none"
                    data-discover="true"
                    href="/binary-tree"
                  >
                    BFS and DFS Traversal
                  </a>
                </ul>
              </div>
            </li>
            <li
              class="ps-menuitem-root ps-submenu-root css-16jesut"
            >
              <a
                class="ps-menu-button"
                data-testid="ps-menu-button-test-id"
                tabindex="0"
              >
                <span
                  class="ps-menu-label css-12w9als"
                >
                  Cryptography
                </span>
                <span
                  class="ps-submenu-expand-icon css-1cuxlhl"
                >
                  <span
                    class="css-honxw6"
                  />
                </span>
              </a>
              <div
                class="ps-submenu-content css-18unl23"
                data-testid="ps-submenu-content-test-id"
              >
                <ul
                  class="css-ewdv3l"
                >
                  <a
                    class="d-block text-center p-1 text-decoration-none"
                    data-discover="true"
                    href="/caesar-cipher"
                  >
                    Caesar Cipher
                  </a>
                </ul>
              </div>
            </li>
            <li
              class="ps-menuitem-root ps-submenu-root css-16jesut"
            >
              <a
                class="ps-menu-button"
                data-testid="ps-menu-button-test-id"
                tabindex="0"
              >
                <span
                  class="ps-menu-label css-12w9als"
                >
                  Computer Graphics
                </span>
                <span
                  class="ps-submenu-expand-icon css-1cuxlhl"
                >
                  <span
                    class="css-honxw6"
                  />
                </span>
              </a>
              <div
                class="ps-submenu-content css-18unl23"
                data-testid="ps-submenu-content-test-id"
              >
                <ul
                  class="css-ewdv3l"
                >
                  <a
                    class="d-block text-center p-1 text-decoration-none"
                    data-discover="true"
                    href="/flood-fill"
                  >
                    Flood Fill
                  </a>
                </ul>
              </div>
            </li>
          </ul>
        </nav>
      </div>
    </aside>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`should render SubMenu component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="ps-menu-root css-vj11vy"
      >
        <ul
          class="css-ewdv3l"
        >
          <li
            class="ps-menuitem-root ps-submenu-root css-16jesut"
          >
            <a
              class="ps-menu-button"
              data-testid="ps-menu-button-test-id"
              tabindex="0"
            >
              <span
                class="ps-menu-label css-12w9als"
              >
                SubMenuTitle
              </span>
              <span
                class="ps-submenu-expand-icon css-1cuxlhl"
              >
                <span
                  class="css-honxw6"
                />
              </span>
            </a>
            <div
              class="ps-submenu-content css-18unl23"
              data-testid="ps-submenu-content-test-id"
            >
              <ul
                class="css-ewdv3l"
              >
                <a
                  class="d-block text-center p-1 text-decoration-none"
                  data-discover="true"
                  href="/"
                >
                  Title
                </a>
              </ul>
            </div>
          </li>
        </ul>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="ps-menu-root css-vj11vy"
    >
      <ul
        class="css-ewdv3l"
      >
        <li
          class="ps-menuitem-root ps-submenu-root css-16jesut"
        >
          <a
            class="ps-menu-button"
            data-testid="ps-menu-button-test-id"
            tabindex="0"
          >
            <span
              class="ps-menu-label css-12w9als"
            >
              SubMenuTitle
            </span>
            <span
              class="ps-submenu-expand-icon css-1cuxlhl"
            >
              <span
                class="css-honxw6"
              />
            </span>
          </a>
          <div
            class="ps-submenu-content css-18unl23"
            data-testid="ps-submenu-content-test-id"
          >
            <ul
              class="css-ewdv3l"
            >
              <a
                class="d-block text-center p-1 text-decoration-none"
                data-discover="true"
                href="/"
              >
                Title
              </a>
            </ul>
          </div>
        </li>
      </ul>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
